### DISCUSSION TITLE
Hook for Fully Formed Checkpoints

### DISCUSSION THREAD

**Discussion by dcharatan** [Lightning-AI/pytorch-lightning#11704] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
I would like to create a hook that automatically uploads checkpoints to the cloud (e.g., AWS, Azure) when they're created. I tried using `on_save_checkpoint` roughly like this:

```
def on_save_checkpoint(self, trainer: pl.Trainer, pl_module: pl.LightningModule, checkpoint: Dict[str, Any]) -> dict:
    checkpoint_bytes = io.BytesIO()
    torch.save(checkpoint, checkpoint_bytes)
    # Upload the BytesIO somewhere...
```

However, states for optimizers, schedulers, AMP, etc. are added after `on_save_checkpoint` hooks are called. Is there an elegant way to create a hook that receives the **fully formed checkpoint state**?

**Comment by rohitgr7** [comment, marked as answer]
hey @dcharatan !

I'd rather suggest using the [remote filesystems](https://pytorch-lightning.readthedocs.io/en/latest/common/remote_fs.html#remote-filesystems). You can also specify the remote path inside `ModelCheckpoint`.

or use [CheckpointIO](https://pytorch-lightning.readthedocs.io/en/latest/common/checkpointing.html#custom-checkpoint-io-plugin) plugin.

**Reply by dcharatan** [reply]
Thanks, this seems like exactly what I was looking for!
